File: Sources/ClientRuntime/Plugins/DefaultClientPlugin.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

import struct SmithyRetries.DefaultRetryStrategy

public class DefaultClientPlugin: Plugin {

    public init() {}

    public func configureClient<Config: ClientConfiguration>(clientConfiguration: inout Config) async throws {
        guard var config = clientConfiguration as? any DefaultClientConfiguration else { return }
        config.retryStrategyOptions =
            DefaultSDKRuntimeConfiguration<DefaultRetryStrategy, DefaultRetryErrorInfoProvider>
                .defaultRetryStrategyOptions

        guard var config2 = config as? any DefaultHttpClientConfiguration else { return }
        let httpClientConfiguration =
            DefaultSDKRuntimeConfiguration<DefaultRetryStrategy, DefaultRetryErrorInfoProvider>
                .defaultHttpClientConfiguration
        config2.httpClientConfiguration = httpClientConfiguration
        config2.httpClientEngine =
            DefaultSDKRuntimeConfiguration<DefaultRetryStrategy, DefaultRetryErrorInfoProvider>
                .makeClient(httpClientConfiguration: httpClientConfiguration)

        guard let modifiedConfig = config2 as? Config else { return }
        clientConfiguration = modifiedConfig
    }
}

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/plugins/DefaultClientPlugin.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

package software.amazon.smithy.swift.codegen.integration.plugins

import software.amazon.smithy.codegen.core.Symbol
import software.amazon.smithy.swift.codegen.SwiftDependency
import software.amazon.smithy.swift.codegen.integration.Plugin
import software.amazon.smithy.swift.codegen.model.buildSymbol
import software.amazon.smithy.swift.codegen.swiftmodules.ClientRuntimeTypes

internal class DefaultClientPlugin : Plugin {
    override val className: Symbol = ClientRuntimeTypes.Core.DefaultClientPlugin

    override val isDefault: Boolean
        get() = true
}
